app/assets/javascripts/discourse/app/lib/hashtag-types/base.js

```diff
@@ -20,4 +29,44 @@ export default class HashtagTypeBase {
   generateIconHTML() {
     throw "not implemented";
   }
+
+  async _load() {
+    // Convert sets to arrays
+    const data = Object.fromEntries(
+      Object.entries(HashtagTypeBase.loadingIds).map(([type, ids]) => [
+        type,
+        [...ids],
+      ])
+    );
+
+    HashtagTypeBase.loadingIds = {};
+
+    const hashtags = await ajax("/hashtags/by-ids", { data });
+    const typeClasses = getHashtagTypeClasses();
+    Object.entries(typeClasses).forEach(([type, typeClass]) =>
+      hashtags[type]?.forEach((hashtag) => typeClass.onLoad(hashtag))
+    );
+  }
+
+  isLoaded(id) {
+    return (
+      !this.site.lazy_load_categories || this.loadedIds.has(parseInt(id, 10))
```

Comment:
Also `loadedIds` isn't actually modified anywhere that I can see, so we do repetitive loading of hashtags that have already been loaded. Also maybe a `Map` with a `Set` inside each hash type key would be better for this? Otherwise `loadedIds` considers a channel ID the same as a category ID if they happen to match.

Reply:
`loadedIds` is per each hash type because it is initialized in the constructor.